### DISCUSSION TITLE
On , Off button for the allsky heater?

### DISCUSSION THREAD

**Discussion by myk68** [aaronwmorris/indi-allsky#1138] [Ideas]
Hi,
Just installed this after finding that it will allow me to use my ASI120mc.  Moved over from the allskyteam option.

I've also installed the Allsky heater from [https://www.instructables.com/Raspberry-Pi-Dew-Heater-for-All-sky-Camera/](https://www.instructables.com/Raspberry-Pi-Dew-Heater-for-All-sky-Camera/).
This is literally just now and found myself wondering if a on-off button could be added to indi-allsky.

Maybe an indicator to show if the heater is on or off, the dome temp figure. 

I know you have a lot of requests but couldn't find a request for this type of thing and thought Id put the Idea up.

Mike

**Comment by aaronwmorris** [comment]
I have thought about this, but I have not come up with a good way to add such a button in the web interface.

One of the major differences between TJ's allsky and indi-allsky is there is no assumption that the software is running on a Raspberry Pi.  Many people run indi-allsky on mini-pcs like NUCs.  Both of my operational cameras are running on RockPi and a Le Potato board.

Any way I add this would have to be either hardware agnostic or added in a way that would not prevent the main program from running if it is not operating on a Raspberry Pi.

**Reply by caphector** [reply]
Could you let the user trigger a custom script? Then the switch is left as an exercise to the user. I'd considered turning mine on simply in response to sunset.

**Reply by aaronwmorris** [reply]
I will be looking into adding some controls that could support multiple types of single board computers.  Possibly to perform some basic focus control with a stepper motor.  Controlling other I/O pins would not be extremely difficult in that case.  I will be looking into it soon.

**Reply by astrotater** [reply]
@aaronwmorris I've been thinking about this. 
Between keeping the program hardware agnostic, and I saw that the Raspberry Pi 5 doesn't have PWM functionality yet, what about using a UART from any indi-allsky board (be it RPi5, Orange Pi, etc) to a slave MCU that is connected to the PWM case fan and PWM dew heater? 
I'm working on an expansion I/O board, my current version is just using a 2N2222 transistor to either turn case fan on or off and dew heater on/off. My 80mm fan draws around 230mA and the dew heater around 260 mA, so the 2N2222 won't be overloaded at that current draw. 
Would a universal UART slave MCU be a workaround? 

I could be going about this the wrong way, but we could use an ATTiny85 and setup serial communication on it using pins 2 and 3, leaving its PWM pins 5 and 6 to PWM control a fan and dew heater. I would think any device could then be setup to use UART communication to that ATTiny85 that is just listening for a command to send a PWM signal to pin 5 or 6. 

Also if you wanted to make a universal basic focus control with a stepper, an ATTiny84 should have enough pins to support a PWM case fan, PWM dew heater, and a ULN2003. 

> I have thought about this, but I have not come up with a good way to add such a button in the web interface.
> 
> One of the major differences between TJ's allsky and indi-allsky is there is no assumption that the software is running on a Raspberry Pi. Many people run indi-allsky on mini-pcs like NUCs. Both of my operational cameras are running on RockPi and a Le Potato board.
> 
> Any way I add this would have to be either hardware agnostic or added in a way that would not prevent the main program from running if it is not operating on a Raspberry Pi.

**Comment by astrotater** [comment]
@aaronwmorris 
This is a quick and dirty example of what I was thinking. I would need to cut a PCB and run some tests as I'm sure this needs cleaning up, but it gets the point across. 

ATTiny firmware: 

```
#include <SoftwareSerial.h>
#include <avr/io.h>

const int Rx = 3; 
const int Tx = 4; 
const int dewHeaterPin = 6;
const int fanPin = 7;
const int IN1 = 10;
const int IN2 = 11;
const int IN3 = 12;
const int IN4 = 13;

SoftwareSerial mySerial(Rx, Tx);

int dewHeaterDutyCycle = 0;
int fanDutyCycle = 0;
int stepperStep = 0;
bool stepperDirection = true; // true for forward, false for reverse

unsigned long previousMillis = 0;
const long stepInterval = 10; 

void setPWM(int pin, int dutyCycle) {
    if (dutyCycle < 0) dutyCycle = 0;
    if (dutyCycle > 255) dutyCycle = 255;
    analogWrite(pin, dutyCycle);
}

void controlStepper() {
    static int currentStep = 0;
    unsigned long currentMillis = millis();
    
    if (currentMillis - previousMillis >= stepInterval) {
        previousMillis = currentMillis;

        const int stepSequence[4][4] = {
            {1, 0, 0, 0},
            {0, 1, 0, 0},
            {0, 0, 1, 0},
            {0, 0, 0, 1}
        };
        
        int adjustedStep = stepperDirection ? currentStep : (3 - currentStep);
        
        digitalWrite(IN1, stepSequence[adjustedStep][0]);
        digitalWrite(IN2, stepSequence[adjustedStep][1]);
        digitalWrite(IN3, stepSequence[adjustedStep][2]);
        digitalWrite(IN4, stepSequence[adjustedStep][3]);
        
        currentStep = (currentStep + 1) % 4;
    }
}

void setup() {
    pinMode(Rx, INPUT);
    pinMode(Tx, OUTPUT);
    mySerial.begin(9600);

    pinMode(dewHeaterPin, OUTPUT);
    pinMode(fanPin, OUTPUT);
    pinMode(IN1, OUTPUT);
    pinMode(IN2, OUTPUT);
    pinMode(IN3, OUTPUT);
    pinMode(IN4, OUTPUT);

    setPWM(dewHeaterPin, 0);
    setPWM(fanPin, 0);
}

void loop() {
    if (mySerial.available() > 0) {
        String command = mySerial.readStringUntil('\n');
        char cmdType = command.charAt(0);
        int value = command.substring(1).toInt();

        switch (cmdType) {
            case 'H': 
                dewHeaterDutyCycle = map(value, 0, 100, 0, 255);
                setPWM(dewHeaterPin, dewHeaterDutyCycle);
                break;
            case 'F': 
                fanDutyCycle = map(value, 0, 100, 0, 255);
                setPWM(fanPin, fanDutyCycle);
                break;
            case 'S':
                stepperDirection = (value > 0);
                stepperStep = abs(value);
                break;
            default:
                break;
        }
    }

    if (stepperStep > 0) {
        controlStepper();
        stepperStep--;
    }
}
```

Sample Raspberry Pi script to receive the UART commands. I think it should work for any machine running Linux that can connect to the slave board I'm working on as long as it connects via UART. That would make PWM fans and dew heaters and stepper motor focusing completely universal. 

```
import serial
import time

SERIAL_PORT = '/dev/ttyS0' 
BAUD_RATE = 9600

def send_command(command):
    with serial.Serial(SERIAL_PORT, BAUD_RATE, timeout=1) as ser:
        ser.write((command + '\n').encode())
        time.sleep(0.1)  

def set_dew_heater(duty_cycle):
    if 0 <= duty_cycle <= 100:
        send_command(f'H{duty_cycle}')
    else:
        print("Duty cycle must be between 0 and 100")

def set_fan(duty_cycle):
    if 0 <= duty_cycle <= 100:
        send_command(f'F{duty_cycle}')
    else:
        print("Duty cycle must be between 0 and 100")

def move_stepper(steps):
    send_command(f'S{steps}')

if __name__ == "__main__":
        choice = input("Enter your choice: ")
        
        if choice == '1':
            duty_cycle = int(input("Enter dew heater duty cycle (0-100): "))
            set_dew_heater(duty_cycle)
        elif choice == '2':
            duty_cycle = int(input("Enter fan duty cycle (0-100): "))
            set_fan(duty_cycle)
        elif choice == '3':
            steps = int(input("Enter number of steps for the stepper motor (positive for one direction, negative for the other): "))
            move_stepper(steps)
        elif choice == '4':
            break
        else:
            print("Invalid choice. Please try again.")
        
        time.sleep(1)
```

        choice = input("Enter your choice: ")
        
        if choice == '1':
            duty_cycle = int(input("Enter dew heater duty cycle (0-100): "))
            set_dew_heater(duty_cycle)
        elif choice == '2':
            duty_cycle = int(input("Enter fan duty cycle (0-100): "))
            set_fan(duty_cycle)
        elif choice == '3':
            steps = int(input("Enter number of steps for the stepper motor (positive for one direction, negative for the other): "))
            move_stepper(steps)
        elif choice == '4':
            break
        else:
            print("Invalid choice. Please try again.")
            
The above code needs to be replaced with whatever commands indi-allsky uses to control the dew heater, fan, and stepper.

**Reply by aaronwmorris** [reply]
This would be trivial to implement, let me think about this over the next few days.

**Reply by astrotater** [reply]
@aaronwmorris Sounds good. If you want to give it a go, send me a python script that has the basic commands that you would want transmitted over UART. I can load the sample script on a test Pi Zero and wire it to the slave board just for trial and error purposes. 
Like what conditions would send a specific PWM signal to the dew heater and fan. 
And what commands you want sent for the stepper. 

Then I can work on firmware for the slave UART board. 
Not sure what timeframe I'm looking at to get it made on my CNC (its down right now, waiting on a new part from Makera), I'm working 60-70 hours this current week and next week, and heading out of town to the beach the following week. 
I could whip something up and have JLCPCB make it and it will be here when I get home.

**Reply by aaronwmorris** [reply]
I am okay with just using the commands you have currently.  Terseness makes sense for an embedded platform.

indi-allsky already has a system for managing the devices and does not really care how it is done, the class for managing the device performs all of the translation between what is requested by indi-allsky and how that request should be interpreted for the device.

**Reply by aaronwmorris** [reply]
I went ahead and added the initial classes for the dew heater and fan support.  #1345   The "pin" for the dew heater and fan will be the name of the serial port (without the /dev/ prefix).

I will add the focus controller support in the next few days.

**Reply by astrotater** [reply]
> I went ahead and added the initial classes for the dew heater and fan support. #1345 The "pin" for the dew heater and fan will be the name of the serial port (without the /dev/ prefix).
> 
> I will add the focus controller support in the next few days.

Ok, I'll get started on a PCB expansion board outline tonight or tomorrow night.

**Comment by astrotater** [comment]
In theory, this works on paper. I'll try and cut it next week and test it out. I need to see if Amazon has all the parts I need to build it and test it. 
12-24v input, separate voltage regulator circuits: 12v for fan and dew heater, 5v for the 2B8YJ-48 stepper connected to the ULN2003 circuit. 
The ATTiny84 gets +5v from the UART connection. I read some folks had issues with the ATTiny if the power supply was separate from the UART connection. I didn't feel like testing it out and then troubleshooting it and ending up back where I am right now, so I'm just starting the circuit off this way. 

![allsky_universal_output_circuit](https://github.com/aaronwmorris/indi-allsky/assets/171597037/76adb7d4-b984-48c2-92c4-f2222caf4385)


Here's the updated ATTiny84 firmware. I suspect this board and code could be universal (assuming it works in real life as well as it does on paper). The AllSkyeye guys on Windows units could use USB->UART and a simple Python program to control it. Same for allsky from TJ's old project, should be easy for them to implement as well. I don't really know what they have going on for fans/dew_heat/focus_motor. 
```
#include <SoftwareSerial.h>
#include <avr/io.h>
#include <avr/interrupt.h>

const int dewHeaterPin = 7; 
const int fanPin = 6;   
const int IN1 = 2;   
const int IN2 = 3;   
const int IN3 = 5;   
const int IN4 = 10;    
const int Rx = 12;     
const int Tx = 11;      

SoftwareSerial mySerial(Rx, Tx);

volatile int dewHeaterDutyCycle = 0;
volatile int fanDutyCycle = 0;
volatile int stepperStep = 0;
volatile bool stepperDirection = true; // true for forward, false for reverse

volatile unsigned long previousMillis = 0;
const long stepInterval = 10; 

void setPWM(int pin, int dutyCycle) {
    dutyCycle = constrain(dutyCycle, 0, 255);
    analogWrite(pin, dutyCycle);
}

void controlStepper() {
    static int currentStep = 0;
    unsigned long currentMillis = millis();
    
    if (currentMillis - previousMillis >= stepInterval) {
        previousMillis = currentMillis;

        const int stepSequence[4][4] = {
            {1, 0, 0, 0},
            {0, 1, 0, 0},
            {0, 0, 1, 0},
            {0, 0, 0, 1}
        };
        
        int adjustedStep = stepperDirection ? currentStep : (3 - currentStep);
        
        digitalWrite(IN1, stepSequence[adjustedStep][0]);
        digitalWrite(IN2, stepSequence[adjustedStep][1]);
        digitalWrite(IN3, stepSequence[adjustedStep][2]);
        digitalWrite(IN4, stepSequence[adjustedStep][3]);
        
        currentStep = (currentStep + 1) % 4;
    }
}

volatile unsigned long timer1_millis = 0;

ISR(TIMER1_COMPA_vect) {
    timer1_millis++;
}

unsigned long millis() {
    unsigned long millis_return;

    cli(); 
    millis_return = timer1_millis;
    sei(); 

    return millis_return;
}

void setupTimer1() {
    TCCR1A = 0;
    TCCR1B = (1 << WGM12) | (1 << CS11); 
    OCR1A = 1999; 

    TIMSK1 |= (1 << OCIE1A); 
    sei(); 
}

void setupPWM() {
    pinMode(fanPin, OUTPUT);
    pinMode(dewHeaterPin, OUTPUT);

    TCCR0A = 0;
    TCCR0B = 0;

    TCCR0A |= (1 << WGM01) | (1 << WGM00);
    TCCR0B |= (1 << WGM02);

    TCCR0B |= (1 << CS00);

    OCR0A = 319; // 8MHz / (1 * (1 + 319)) = 25kHz

    OCR0B = 0;

    TCCR0A |= (1 << COM0A1) | (1 << COM0B1);
}

void setup() {
    setupTimer1();
    setupPWM();
    pinMode(Rx, INPUT);
    pinMode(Tx, OUTPUT);
    mySerial.begin(9600); 

    pinMode(IN1, OUTPUT);
    pinMode(IN2, OUTPUT);
    pinMode(IN3, OUTPUT);
    pinMode(IN4, OUTPUT);
}

void loop() {
    if (mySerial.available() > 0) {
        char cmdType = mySerial.read();
        int value = mySerial.parseInt();

        switch (cmdType) {
            case 'H': // Dew heater command
                dewHeaterDutyCycle = map(value, 0, 100, 0, 255);
                setPWM(dewHeaterPin, dewHeaterDutyCycle);
                break;
            case 'F': // Fan command
                fanDutyCycle = map(value, 0, 100, 0, 255);
                setPWM(fanPin, fanDutyCycle);
                break;
            case 'S': // Stepper motor command
                stepperDirection = (value > 0);
                stepperStep = abs(value);
                break;
            default:
                break;
        }
    }

    if (stepperStep > 0) {
        controlStepper();
        stepperStep--;
    }
}

void analogWrite(uint8_t pin, uint8_t value) {
    if (pin == fanPin) {
        OCR0A = map(value, 0, 255, 0, 319); 
    } else if (pin == dewHeaterPin) {
        OCR0B = map(value, 0, 255, 0, 319); 
    }
}

void pinMode(uint8_t pin, uint8_t mode) {
    if (mode == OUTPUT) {
        if (pin == fanPin) {
            DDRB |= (1 << DDB2);
        } else if (pin == dewHeaterPin) {
            DDRB |= (1 << DDB3);
        } else if (pin == IN1) {
            DDRA |= (1 << DDA0); 
        } else if (pin == IN2) {
            DDRA |= (1 << DDA1); 
        } else if (pin == IN3) {
            DDRA |= (1 << DDA2); 
        } else if (pin == IN4) {
            DDRA |= (1 << DDA4);
        } else if (pin == Rx) {
            DDRA &= ~(1 << DDA1); 
        } else if (pin == Tx) {
            DDRA |= (1 << DDA2); 
        }
    } else if (mode == INPUT) {
        if (pin == Rx) {
            DDRA &= ~(1 << DDA1); 
        }
    }
}

void digitalWrite(uint8_t pin, uint8_t value) {
    if (pin == IN1) {
        if (value == HIGH) {
            PORTA |= (1 << PA0);
        } else {
            PORTA &= ~(1 << PA0);
        }
    } else if (pin == IN2) {
        if (value == HIGH) {
            PORTA |= (1 << PA1);
        } else {
            PORTA &= ~(1 << PA1);
        }
    } else if (pin == IN3) {
        if (value == HIGH) {
            PORTA |= (1 << PA2);
        } else {
            PORTA &= ~(1 << PA2);
        }
    } else if (pin == IN4) {
        if (value == HIGH) {
            PORTA |= (1 << PA4);
        } else {
            PORTA &= ~(1 << PA4);
        }
    } else if (pin == Tx) {
        if (value == HIGH) {
            PORTA |= (1 << PA2);
        } else {
            PORTA &= ~(1 << PA2);
        }
    }
}

```

**Reply by aaronwmorris** [reply]
Merged #1346 for the initial serial focuser support.

You will probably want to add a voltage divider on the TX pin (11) if this is intended to be directly connected to the serial pins of a SBC (like Raspi) since the pins are not 5v tolerant.  You would only need the divider on the TX pin of the ATtiny.  This would be compatible with both 5v and 3.3v systems.  You could also do a bidirectional level shifter, but serial comms are so slow (relatively) the voltage divider is usually sufficient.

**Comment by astrotater** [comment]
I'll test this sometime when I get a chance. 
If I get some errors, I'll give the level shifter a shot. 


![Screenshot 2024-06-23 011853](https://github.com/aaronwmorris/indi-allsky/assets/171597037/52bf2ef9-6d82-485e-9ff1-7bf7babc3f1c)

**Reply by aaronwmorris** [reply]
That is amazingly quick.  :-)

**Reply by astrotater** [reply]
I just hope it works!
I’m having the prototype made at JLCPCB, I don’t have time this week to cut my own on the CNC and I’m leaving town next Monday. So I probably won’t be able to test the boards UART functionality for roughly 2-2.5 weeks. 

When I get it tested and if it works, I’ll send you one.

**Comment by astrotater** [comment]
@aaronwmorris got the boards from JLCPCB, waiting on some transistors and FETs. 
Question, as I'm putting it together, what are your thoughts on switching from an attiny84 to an Arduino Nano and connecting it to the indi-allsky over USB? And adding I2C to the PCB. 
People can add whatever I2C devices they want, and pass the I2C signals to the indi-allsky. StickPCs and NUCs can also gain I2C interfaces if I use an Arduino Nano instead of attiny84. 

As I work on the attiny84, it's a PITA, and I'm not sure it would be easy enough to implement for users with no experience. Whereas an Arduino Nano, almost anyone could upload firmware to that and get the board working. Arduino Nano could also be updated with new firmware from whatever computer (Pi, LePotato, StickPC, NUC, etc) on the fly, an attiny84 can't.

**Reply by aaronwmorris** [reply]
I am not sure I understand how moving to a Nano would enable I2C for PCs like NUCs.  I can totally see how using a Nano would make your development workflow easier.

**Reply by astrotater** [reply]
A Nano has enough pins to connect I2C devices, daisy chained for multiple sensors. 
It could pass I2C sensor data to a NUC over the UART, could it not?

**Reply by aaronwmorris** [reply]
I see, so you are still using the UART for communication, but you can use I2C to communicate to sensors, etc.  That makes perfect sense.

This would require some bidirectional communication via serial (as opposed to just sending commands), but that is not difficult.

**Reply by astrotater** [reply]
Yeah, while I was beginning to wire it all up, my idea was [NUC/Pi/Libre/OrangePi/StickPC/etc] <-USB Serial-> Arduino Nano ->PWM dew heater and fan, I2C BME280/BME680/TSL2591/MLX90614, one wire DS18B20 temp, ULN2003 focusing, etc. 
Basically one USB serial wire to the indi-allsky computer, whatever that may be, and the Arduino can handle all the interfacing with everything else. That would make it truly universal for all sensor types connected to any indi PC, regardless if it has GPIO or not. 
Kinda pointless with a Pi that has GPIO, but useful for all other mini PC types. I think it could potentially make cable runs cleaner with a Pi too.

**Comment by astrotater** [comment]
I haven't forgotten about this. 
Just received the PCBs today from the slow boat from China. 
Getting killed at work, I'm hoping to install the components on the PCB at sometime in the next week and test it out.
